OutOfSchool/OutOfSchool.Tests/IdentityServerTests/AuthControllerTests.cs

```diff
@@ -0,0 +1,299 @@
+                    manager.PasswordSignInAsync(It.IsAny<string>(), It.IsAny<string>(), false, false))
+                .ReturnsAsync(expectedResult.Key);
+            // Act
+            var result = await authController.Login(loginViewModel);
+            // Assert
+            Type type = expectedResult.Value;
+            Assert.IsInstanceOf(type, result);
+        }
+
+        [TestCase("")]
+        [TestCase("Return url")]
+        public async Task RegisterWithReturnUrl(string returnUrl)
+        {
+            // Arrange 
+            var authController = CreateAuthController;
+            mockRoleManager.Setup(manager => manager.Roles)
+                .Returns(new List<IdentityRole>().AsQueryable());
+            // Act
+            var result = authController.Register(returnUrl);
+            // Assert
+            Assert.IsInstanceOf<ViewResult>(result);
+        }
+
+        [Test]
+        public async Task RegisterWithVMWithModelError()
+        {
+            // Arrange 
+            var authController = CreateAuthController;
+            mockRoleManager.Setup(manager => manager.Roles)
+                .Returns(new List<IdentityRole>().AsQueryable());
+            authController.ModelState.AddModelError("", "Message");
+            // Act
+            var result = await authController.Register(new RegisterViewModel());
+            // Assert
+            Assert.IsInstanceOf<ViewResult>(result);
+        }
+
+        [TestCaseSource(nameof(RegisterViewModelsTestData))]
+        public async Task RegisterWithVMWithoutModelError(RegisterViewModel viewModel)
+        {
+            // Arrange 
+            var authController = CreateAuthController;
+            mockRoleManager.Setup(manager => manager.Roles)
+                .Returns(
+                    new List<IdentityRole>() {new IdentityRole("testrole") {Id = viewModel.UserRoleId}}.AsQueryable());
+            mockUserManager.Setup(manager => manager.CreateAsync(It.IsAny<User>(), viewModel.Password))
+                .Returns(Task.FromResult<IdentityResult>(IdentityResult.Success));
+            mockUserManager.Setup(manager => manager.AddToRoleAsync(It.IsAny<User>(), It.IsAny<string>()))
+                .Returns(Task.FromResult<IdentityResult>(IdentityResult.Success));
+            mockSignInManager.Setup(manager => manager.SignInAsync(
+                    It.IsAny<User>(), It.IsAny<bool>(), It.IsAny<string>()))
+                .Returns(Task.CompletedTask);
+            // Act
+            var result = await authController.Register(viewModel);
+            // Assert
+            Assert.IsInstanceOf<RedirectResult>(result);
+        }
+
+        [TestCaseSource(nameof(RegisterViewModelsTestData))]
+        public async Task RegisterWithVMAndIdentityError(RegisterViewModel viewModel)
+        {
+            // Arrange 
+            var authController = CreateAuthController;
+            var identityResult = IdentityResult.Failed(new List<IdentityError>()
```

Comment:
if an expression is multiline, adding newline before and after might improve readability